Utils/src/HiddenSectorProducer.cc

```diff
@@ -22,143 +27,390 @@
+std::vector<int> HiddenSectorProducer::matchGenRec(const edm::View<pat::Jet>& jets, const edm::View<reco::GenJet>& gen) const {
+  std::map<double,std::pair<unsigned,unsigned>> matchMap;
+  for(unsigned j = 0; j < jets.size(); ++j){
+    for(unsigned g = 0; g < gen.size(); ++g){
+      matchMap.emplace(std::piecewise_construct, std::forward_as_tuple(reco::deltaR(jets[j],gen[g])), std::forward_as_tuple(j,g));
+    }
+  }
+
+  std::vector<int> genIndex(jets.size(),-1);
+  std::unordered_set<unsigned> j_used, g_used;
+  for(const auto& matchItem: matchMap){
+    unsigned j = matchItem.second.first;
+    unsigned g = matchItem.second.second;
+    if(j_used.find(j)==j_used.end() and g_used.find(g)==g_used.end()){
+      genIndex[j] = g;
+      j_used.insert(j);
+      g_used.insert(g);
+    }
+  }
+
+  return genIndex;
+}
+
 HiddenSectorProducer::HiddenSectorProducer(const edm::ParameterSet& iConfig) :
-	JetTag_(iConfig.getParameter<edm::InputTag>("JetTag")),
-	MetTag_(iConfig.getParameter<edm::InputTag>("MetTag")),
-	GenTag_(iConfig.getParameter<edm::InputTag>("GenTag")),
-	JetTok_(consumes<edm::View<pat::Jet>>(JetTag_)),
-	MetTok_(consumes<edm::View<pat::MET>>(MetTag_)),
-	GenTok_(consumes<edm::View<reco::GenParticle>>(GenTag_)),
-	DarkQuarkID_(iConfig.getParameter<unsigned>("DarkQuarkID")),
-	DarkMediatorID_(iConfig.getParameter<unsigned>("DarkMediatorID"))
+  JetTag_(iConfig.getParameter<edm::InputTag>("JetTag")),
+  MetTag_(iConfig.getParameter<edm::InputTag>("MetTag")),
+  GenMetTag_(iConfig.getParameter<edm::InputTag>("GenMetTag")),
+  GenTag_(iConfig.getParameter<edm::InputTag>("GenTag")),
+  GenJetTag_(iConfig.getParameter<edm::InputTag>("GenJetTag")),
+  JetTok_(consumes<edm::View<pat::Jet>>(JetTag_)),
+  MetTok_(consumes<edm::View<pat::MET>>(MetTag_)),
+  GenMetTok_(consumes<edm::View<reco::GenMET>>(GenMetTag_)),
+  GenTok_(consumes<edm::View<reco::GenParticle>>(GenTag_)),
+  GenJetTok_(consumes<edm::View<reco::GenJet>>(GenJetTag_)),
+  coneSize_(iConfig.getParameter<double>("coneSize"))
 {
-	const auto& ids = iConfig.getParameter<std::vector<unsigned>>("DarkIDs");
-	DarkIDs_.insert(ids.begin(),ids.end());
-
-	produces<double>("MJJ");
-	produces<double>("Mmc");
-	produces<double>("MT");
-	produces<double>("DeltaPhi1");
-	produces<double>("DeltaPhi2");
-	produces<double>("DeltaPhiMin");
-	produces<std::vector<bool>>("isHV");
+  fillSet(DarkMediatorIDs_,"DarkMediatorIDs",iConfig);
+  fillSet(DarkQuarkIDs_,"DarkQuarkIDs",iConfig);
+  fillSet(DarkHadronIDs_,"DarkHadronIDs",iConfig);
+  fillSet(DarkGluonIDs_,"DarkGluonIDs",iConfig);
+  fillSet(DarkStableIDs_,"DarkStableIDs",iConfig);
+  fillSet(SMQuarkIDs_,"SMQuarkIDs",iConfig);
+  //combined list of possible first particles
+  DarkFirstIDs_.insert(DarkMediatorIDs_.begin(),DarkMediatorIDs_.end());
+  DarkFirstIDs_.insert(DarkQuarkIDs_.begin(),DarkQuarkIDs_.end());
+  DarkFirstIDs_.insert(DarkGluonIDs_.begin(),DarkGluonIDs_.end());
+
```

Comment:
add a call to `asymm_mt2_lester_bisect::disableCopyrightMessage();` here to block the unnecessary printout